File: AdventOfCode2020/Day4/Day4a.fs
module Day4a

open System
open System.IO
open System.Collections.Generic

let rec combinePair xs =
    match xs with
    | [] | [_] -> []
    | x1::x2::rest -> (x1, x2) :: (combinePair rest)

let handleLine (current : Dictionary<string, string>) (line : string) =
    let zipped = line.Split(" :".ToCharArray(), StringSplitOptions.RemoveEmptyEntries) |> Array.toList |> combinePair
    zipped |> List.iter (fun (k, v) -> current.Add(k, v))
    current

let rec parseLines passports current lines =
    match lines with
    | [] -> (current :: passports)
    | x::rest ->
        match x with
        | "" -> parseLines (current :: passports) (new Dictionary<string, string>()) rest
        | _ -> parseLines passports (handleLine current x) rest

let isValidPassport (passport : Dictionary<string, string>) =
    passport.Count = 8 || (passport.Count = 7 && not (passport.ContainsKey("cid")))

let solver =
    File.ReadLines "Inputs\\day4.txt"
        |> List.ofSeq
        |> parseLines List.empty<Dictionary<string, string>> (new Dictionary<string, string>())
        |> List.filter isValidPassport
        |> List.length
        |> printfn "%d"

File: AdventOfCode2024/Day4/Day4A.cs
namespace AdventOfCode2024.Day4;

public class Day4A : ISolver
{
    public void Solve()
    {
        var lines = File.ReadAllLines("Inputs\\day4.txt");
        
        var maxX = lines[0].Length;
        var maxY = lines.Length;

        var count = 0;
        for (var y = 0; y < lines.Length; y++)
        {
            for (var x = 0; x < lines[0].Length; x++)
            {
                if (lines[y][x] != 'X')
                {
                    continue;
                }
                
                if (x < maxX - 3)
                {
                    if (lines[y][x + 1] == 'M' && lines[y][x + 2] == 'A' && lines[y][x + 3] == 'S')
                    {
                        count++;
                    }
                    
                    if (y < maxY - 3)
                    {
                        if (lines[y + 1][x + 1] == 'M' && lines[y + 2][x + 2] == 'A' && lines[y + 3][x + 3] == 'S')
                        {
                            count++;
                        }
                    }

                    if (y > 2)
                    {
                        if (lines[y - 1][x + 1] == 'M' && lines[y - 2][x + 2] == 'A' && lines[y - 3][x + 3] == 'S')
                        {
                            count++;
                        }
                    }
                }

                if (x > 2)
                {
                    if (lines[y][x - 1] == 'M' && lines[y][x - 2] == 'A' && lines[y][x - 3] == 'S')
                    {
                        count++;
                    }
                    
                    if (y < maxY - 3)
                    {
                        if (lines[y + 1][x - 1] == 'M' && lines[y + 2][x - 2] == 'A' && lines[y + 3][x - 3] == 'S')
                        {
                            count++;
                        }
                    }

                    if (y > 2)
                    {
                        if (lines[y - 1][x - 1] == 'M' && lines[y - 2][x - 2] == 'A' && lines[y - 3][x - 3] == 'S')
                        {
                            count++;
                        }
                    }
                }

                if (y < maxY - 3)
                {
                    if (lines[y + 1][x] == 'M' && lines[y + 2][x] == 'A' && lines[y + 3][x] == 'S')
                    {
                        count++;
                    }
                }

                if (y > 2)
                {
                    if (lines[y - 1][x] == 'M' && lines[y - 2][x] == 'A' && lines[y - 3][x] == 'S')
                    {
                        count++;
                    }
                }
            }
        }
        
        Console.WriteLine(count);
    }
}